custom_metrics/almanac.js

```diff
@@ -7,37 +7,223 @@
 //
 // 1. The key should be a unique identifier for the metric, if possible, like it's metric ID (eg '01.12').
 // 1a. If the value is used by multiple metrics, the key should be named according to what it's measuring (eg 'link-nodes').
-// 2. The value must evaluate to a simple type like int or string. For complex types (array, obj) use `JSON.stringify`.
-// 2a. If the value requires more than one line of code, evaluate it in an IIFE, eg `(() => { ... })()`. See `link-nodes`.
+// 2. If the value requires more than one line of code, evaluate it in an IIFE, eg `(() => { ... })()`. See `link-nodes`.
 // 3. Test your change by following the instructions at https://github.com/HTTPArchive/almanac.httparchive.org/issues/33#issuecomment-502288773.
 // 4. Submit a PR to update this file.
 
-return {
+function parseNodes(nodes) {
+  var parsedNodes = [];
+  if (nodes) {
+    for (var i = 0, len = nodes.length; i < len; i++) {
+      var node = nodes[i];
+      var attributes = Object.values(node.attributes);
+      var el = {};
+
+      el.tagName = node.tagName.toLowerCase(); // for reference
+      for (var n = 0, len2 = attributes.length; n < len2; n++) {
+        var attribute = attributes[n];
+        el[attribute.name.toLowerCase()] = attribute.value;
+      }
+
+      parsedNodes.push(el);
+    }
+  }
+  return parsedNodes;
+}
+
+return JSON.stringify({
   // Wether the page contains <script type=module>.
   '01.12': document.querySelector('script[type=module]') ? 1 : 0,
   // Wether the page contains <script nomodule>.
   '01.13': document.querySelector('script[nomodule]') ? 1 : 0,
   'link-nodes': (() => {
     // Returns a JSON array of link nodes and their key/value attributes.
-    // Used by 01.14, 01.15, 01.16...
+    // Used by 01.14, 01.15, 01.16, 10.6,  06.46, 12.18
     var nodes = document.querySelectorAll('head link');
-    var linkNodes = [];
+    var linkNodes = parseNodes(nodes);
+
+    return linkNodes;
+  })(),
+  'meta-nodes': (() => {
+    // Returns a JSON array of meta nodes and their key/value attributes.
+    // Used by 10.6, 10.7 (potential: 09.29, 12.5, 04.5)
+    var nodes = document.querySelectorAll('head meta');
+    var metaNodes = parseNodes(nodes);
+
+    return metaNodes;
+  })(),
+  // Extract schema.org elements and finds all @context and @type usage
+  '10.5': (() => {
+    function nestedLookup(items) {
+      var keys = Object.keys(items);
+      for (var i = 0, len = keys.length; i < len; i++) {
+        var item = items[keys[i]];
+        // if array or object, dive into it
+        if (item instanceof Object || item instanceof Array) {
+          nestedLookup(item);
```

Comment:
Hmm it's concerning that this is recursive. We run 5M tests for both desktop and mobile and I worry that O(N^2) algorithms could have an adverse impact on the performance of the crawl. Do you have any anecdotal experience with the size/depth of `items`?

cc @pmeenan for second opinions.

Reply:
I understand. I could add a maxDepth. E.g. 5 would be more than enough for parsing schema.org data.